DTS manifest file part 1

briehl opened the pull request:
This includes the first steps toward having the DTS manifest files as import specifications. It does the following:

1. Adds `.json` to the allowed import spec files, as `DTS_MANIFEST`
2. Maps `DTS_MANIFEST` to a (for now) dummy parser function.
3. Starts up the dummy parser function with basic fail conditions - file not found, file is a directory, file is wrong (basic) format, etc.

Next PR will make the parser do some parsing.

briehl committed:
add .json files (dts import manifest) to import specification list

briehl committed:
add dummy parse_dts_manifest function

briehl committed:
progress on dts_manifest base error tests

briehl committed:
add dummy dts manifest parser and basic fail tests

briehl committed:
add JSON format check and test

briehl committed:
fix test fixture docstrong

briehl committed:
ruff formatting

MrCreosote commented on staging_service/autodetect/Mappings.py:
```diff
@@ -99,7 +102,8 @@ def _add_gzip(extension_list):
     TSV: ["tsv"],  # See Note 1 below
-    CSV: ["csv"],  # See Note 1 below
+    CSV: ["csv"],  # See Note 1 below,
+    DTS_MANIFEST: ["json"],
     JSON: ["json"],
```
DTS_MANIFEST and JSON are both the text `JSON` and so collide in the map

MrCreosote reviewed (commented):
It's been so long I don't really remember how the mapping mechanism works

MrCreosote commented:
Ugh, hit finish review too early. I'm not done

MrCreosote commented on staging_service/autodetect/Mappings.py:
```diff
@@ -7,6 +7,9 @@
 EXCEL = "EXCEL"
 ZIP = "CompressedFileFormatArchive"
 
+# Data Transfer Service Manifest format (which is a specific JSON format)
+DTS_MANIFEST = "JSON"
+
```
These types are just the file types, regardless of the semantics of the type. Just JSON should be used here

briehl replied:
Done.

MrCreosote commented on deployment/conf/supported_apps_w_extensions.json:
```diff
@@ -1122,8 +1123,13 @@
+        {
+          "id": "import_specification",
+          "title": "Import Specification",
+          "app_weight": 1
```
This file is supposed to be autogenerated IIRC but I don't see any code changes that would cause this to be different

MrCreosote replied:
IIRC you need to update the mapping here: https://github.com/kbase/staging_service/blob/d41d38c5851350c231950a52d176f9c20a065516/staging_service/autodetect/GenerateMappings.py#L66

briehl replied:
Well that'd be real cool if that were documented anywhere. I'll add that and fix it.

MrCreosote replied:
https://github.com/kbase/staging_service/issues/96

briehl replied:
Added some documentation to the README. Not perfect, but should be enough for now.

MrCreosote reviewed (changes requested):
Holding the review here as I think the app configuration is incorrect

briehl committed:
fix silliness around dts manifest mapping

briehl committed:
add link from API call to instructions on how to update mappings

briehl commented:
Ok, that was a very silly change I did.
New version:
* Removed that DTS_MANIFEST mapping nonsense, just stick with JSON.
* Ran GenerateMappings.py to update the mapping JSON file. Which really didn't change anything, but we should consider best practice.
* Added some lightweight documentation on the file mapping stuff and how to update it. Should come back to that at some point.

briehl commented:
Trivy's being whiny again: `TOOMANYREQUESTS: retry-after: 483.512µs, allowed: 44000/minute"`

MrCreosote commented on staging_service/autodetect/GenerateMappings.py:
```diff
@@ -120,7 +129,7 @@
     import_specification,
 ]
-file_format_to_app_mapping[JSON] = [escher_map_id]
+file_format_to_app_mapping[JSON] = [escher_map_id, import_specification]
```
Now I'm wondering if we should have a separate type in the dropdown for DTS manifests vs. import specifications. Maybe just `DTS manifest`? That would be forward compatible if we ever wanted to support standard import specifications with JSON input.

briehl replied:
Yeah, I thought of that. I guess that would be a trivial change, but I'm not sure the best thing to call it. I also wonder if it would be confusing to users at all? They'd have to read instructions on the whole DTS process anyway.

briehl replied:
Oh, I take it back, that's not trivial at all. It means a narrative change at minimum, possibly an API change if we want to be really thorough. Maybe a new API endpoint?

MrCreosote replied:
I would make it a query param probably

MrCreosote replied:
Just checking here - you're saying the separate type and query param is the way you're going to go?

MrCreosote replied:
Also, if it's going to work that way, there probably shouldn't be a JSON: dts_parser entry in the parser mapping

briehl replied:
This is how it work right now, as of this PR, and everything else that's in it. A later PR that updates the endpoint to use a query param will probably undo this. Also, the Narrative will need to have a mapping of some sort, which is what this sets up. So it might not be from `.json` -> `Import Specification`, but it'll need to be something similar.

I don't want this to get more complicated and would like to keep that in a single PR with the `bulk_specification` endpoint change.

MrCreosote replied:
Fair enough, as long as the plan is clear

MrCreosote commented on staging_service/autodetect/Mappings.py:
```diff
@@ -99,7 +99,7 @@ def _add_gzip(extension_list):
     # ],
     TSV: ["tsv"],  # See Note 1 below
-    CSV: ["csv"],  # See Note 1 below
+    CSV: ["csv"],  # See Note 1 below,
```
typo?

briehl replied:
fixed

MrCreosote replied:
:+1:

MrCreosote commented on staging_service/import_specifications/individual_parsers.py:
```diff
@@ -316,3 +317,50 @@ def parse_excel(path: Path) -> ParseResults:
+        with open(path, "r") as manifest:
+            manifest_json = json.load(manifest)
+        if not isinstance(manifest_json, dict):
+            errors.append(Error(ErrorType.PARSE_FAIL, "Manifest is not a dictionary", spcsrc))
+
+    except json.JSONDecodeError:
```
I assume line 354 is where the magic occurs

briehl replied:
Yup, next PR. Or the one after.

MrCreosote replied:
:+1:

MrCreosote commented on staging_service/import_specifications/individual_parsers.py:
```diff
@@ -316,3 +317,50 @@ def parse_excel(path: Path) -> ParseResults:
+def parse_dts_manifest(path: Path) -> ParseResults:
+    """
+    Parse the provided DTS manifest file. Expected to be JSON, and will fail otherwise.
+    The manifest should have roughly this format, with expected keys included:
```
If I provide a manifest that only has roughly the format below will it parse? I'd expect not

briehl replied:
In this PR? No, it's mostly a null placeholder to satisfy the mapping because I didn't want to make this one much larger. Basic fails are trapped properly, and a "passing" file returns nothing.

MrCreosote replied:
I'm talking about this line

> The manifest should have roughly this format

Which seems to imply that the parser is smart enough to deal with variations in format, which I'd assume is something you're not interested in doing

briehl replied:
Updated that docstring to be more clear.

MrCreosote replied:
Did you push the commit? I'm not seeing any changes

MrCreosote replied:
:+1:

MrCreosote commented on staging_service/import_specifications/individual_parsers.py:
```diff
@@ -316,3 +317,50 @@ def parse_excel(path: Path) -> ParseResults:
+            }
+        }]
+    }
+    The parameters under the "instructions"."parameters" dictionary are arbitrary keys with
+    arbitrary values and will be returned as-is. They are expected to be PRIMITIVE_TYPEs and
+    will fail otherwise.
```
This seems like it's expanding the import specification return format, which presumably the front end won't understand?

briehl replied:
Nope, it'll return the same format. I'll edit to clarify.

MrCreosote replied:
:+1:

MrCreosote commented on staging_service/import_specifications/individual_parsers.py:
```diff
@@ -316,3 +317,50 @@ def parse_excel(path: Path) -> ParseResults:
+    # dummy for now
+    results = {}
+    try:
+        with open(path, "r") as manifest:
+            manifest_json = json.load(manifest)
+        if not isinstance(manifest_json, dict):
+            errors.append(Error(ErrorType.PARSE_FAIL, "Manifest is not a dictionary", spcsrc))
```
You might want to consider jsonschema for checking the structure

briehl replied:
Next PR

MrCreosote replied:
:+1:

briehl committed:
fix typo

briehl committed:
update docstrings, add TODOS

briehl committed:
remove 'roughly'

MrCreosote commented on tests/import_specifications/test_data/manifest_small.json:
```diff
@@ -0,0 +1,122 @@
+{
```
This doesn't seem to match the documentation in individual_parsers but I assume the format is still changing

MrCreosote reviewed (approved):
LGTM given the caveats re reworking this have have a specific DTS manifest type
